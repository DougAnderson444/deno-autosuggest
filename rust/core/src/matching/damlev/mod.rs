mod matrix;

use std::f64;
use fnv::{FnvHashMap as HashMap};
use std::cell::RefCell;
use crate::lang::CharClass;
use crate::tokenization::{Word, WordView};
use matrix::DistMatrix;


const DEFAULT_CAPACITY: usize = 20;

const COST_TRANS:     f64 = 0.5;
const COST_DOUBLE:    f64 = 0.5;
const COST_VOWEL:     f64 = 0.5;
const COST_NOTALPHA:  f64 = 0.5;
const COST_CONSONANT: f64 = 1.0;
const COST_DEFAULT:   f64 = 1.0;


pub struct DamerauLevenshtein {
    pub dists: RefCell<DistMatrix>,
    last_i1: RefCell<HashMap<char, usize>>,
    costs1:  RefCell<Vec<f64>>,
    costs2:  RefCell<Vec<f64>>,
}


impl DamerauLevenshtein {
    pub fn new() -> Self {
        let dists   = RefCell::new(DistMatrix::new(DEFAULT_CAPACITY + 2));
        let last_i1 = RefCell::new(HashMap::with_capacity_and_hasher(DEFAULT_CAPACITY, Default::default()));
        let costs1  = RefCell::new(Vec::with_capacity(DEFAULT_CAPACITY));
        let costs2  = RefCell::new(Vec::with_capacity(DEFAULT_CAPACITY));
        Self { dists, last_i1, costs1, costs2 }
    }

    fn get_cost(class: &CharClass) -> f64 {
        match class {
            CharClass::Consonant => COST_CONSONANT,
            CharClass::Vowel     => COST_VOWEL,
            CharClass::NotAlpha  => COST_NOTALPHA,
            _                    => COST_DEFAULT,
        }
    }

    pub fn distance(&self, word1: &WordView, word2: &WordView) -> f64 {
        let chars1 = word1.chars();
        let chars2 = word2.chars();

        let costs1 = &mut *self.costs1.borrow_mut();
        let costs2 = &mut *self.costs2.borrow_mut();
        costs1.clear();
        costs2.clear();
        costs1.extend(word1.classes().iter().map(Self::get_cost));
        costs2.extend(word2.classes().iter().map(Self::get_cost));

        let dists = &mut *self.dists.borrow_mut();
        dists.prepare(&costs1, &costs2);

        let last_i1 = &mut *self.last_i1.borrow_mut();
        last_i1.clear();

        for (i1, &ch1) in chars1.iter().enumerate() {
            let mut l2 = 0;

            let cost1        = unsafe { *costs1.get_unchecked(i1) };
            let double1      = i1 > 0 && ch1 == unsafe { *chars1.get_unchecked(i1 - 1) };
            let cost_double1 = if double1 { COST_DOUBLE } else { COST_DEFAULT };
            let cost_del     = fmin(cost1, cost_double1);

            for (i2, &ch2) in chars2.iter().enumerate() {
                let l1 = *last_i1.get(&ch2).unwrap_or(&0);

                let cost2        = unsafe { *costs2.get_unchecked(i2) };
                let double2      = i2 > 0 && ch2 == unsafe { *chars2.get_unchecked(i2 - 1) };
                let cost_double2 = if double2 { COST_DOUBLE } else { COST_DEFAULT };
                let cost_add     = fmin(cost2, cost_double2);
                let cost_sub     = if ch1 == ch2 { 0.0 } else { fmax(cost1, cost2) };
                let cost_trans   = COST_TRANS * ((i1 - l1) + (i2 - l2) + 1) as f64;

                let dist_add   = cost_add   + unsafe { dists.get_unchecked(i1 + 2, i2 + 1) };
                let dist_del   = cost_del   + unsafe { dists.get_unchecked(i1 + 1, i2 + 2) };
                let dist_sub   = cost_sub   + unsafe { dists.get_unchecked(i1 + 1, i2 + 1) };
                let dist_trans = cost_trans + unsafe { dists.get_unchecked(l1, l2) };
                let dist       = fmin4(dist_add, dist_del, dist_sub, dist_trans);

                unsafe {
                    dists.set_unchecked(i1 + 2, i2 + 2, dist);
                }

                if ch1 == ch2 { l2 = i2 + 1; }
            }
            last_i1.insert(ch1, i1 + 1);
        }

        unsafe { dists.get_unchecked(word1.len() + 1, word2.len() + 1) }
    }
}


#[inline]
fn fmin4(x1: f64, x2: f64, x3: f64, x4: f64) -> f64 {
    let mut min = x1;
    if x2 < min { min = x2; }
    if x3 < min { min = x3; }
    if x4 < min { min = x4; }
    min
}

#[inline]
fn fmin(x1: f64, x2: f64) -> f64 {
    if x1 < x2 { x1 } else { x2 }
}

#[inline]
fn fmax(x1: f64, x2: f64) -> f64 {
    if x1 > x2 { x1 } else { x2 }
}


#[cfg(test)]
mod tests {
    use crate::lang::lang_english;
    use crate::tokenization::Text;
    use super::DamerauLevenshtein;


    #[test]
    fn equality() {
        let damlev = DamerauLevenshtein::new();
        let sample = [
            Text::from_str(""),
            Text::from_str("a"),
            Text::from_str("ab"),
            Text::from_str("abc"),
        ];
        for text in sample.iter() {
            assert_eq!(damlev.distance(&text.view(0), &text.view(0)), 0.0);
        }
    }

    #[test]
    fn prefix() {
        let damlev = DamerauLevenshtein::new();
        let sample = [
            (0.0, Text::from_str("abc"), Text::from_str("abc")),
            (1.0, Text::from_str("abc"), Text::from_str("ab")),
            (2.0, Text::from_str("abc"), Text::from_str("a")),
            (3.0, Text::from_str("abc"), Text::from_str("")),
        ];
        for (d, t1, t2) in sample.iter() {
            assert_eq!(damlev.distance(&t1.view(0), &t2.view(0)), *d);
            assert_eq!(damlev.distance(&t2.view(0), &t1.view(0)), *d);
        }
    }

    #[test]
    fn add_del_continuous() {
        let damlev = DamerauLevenshtein::new();
        let sample = [
            (1.0, Text::from_str("abc"), Text::from_str("xabc")),
            (2.0, Text::from_str("abc"), Text::from_str("xyabc")),
            (3.0, Text::from_str("abc"), Text::from_str("xyzabc")),

            (1.0, Text::from_str("abc"), Text::from_str("axbc")),
            (2.0, Text::from_str("abc"), Text::from_str("axybc")),
            (3.0, Text::from_str("abc"), Text::from_str("axyzbc")),

            (1.0, Text::from_str("abc"), Text::from_str("abcx")),
            (2.0, Text::from_str("abc"), Text::from_str("abcxy")),
            (3.0, Text::from_str("abc"), Text::from_str("abcxyz")),
        ];
        for (d, t1, t2) in sample.iter() {
            assert_eq!(damlev.distance(&t1.view(0), &t2.view(0)), *d);
            assert_eq!(damlev.distance(&t2.view(0), &t1.view(0)), *d);
        }
    }

    #[test]
    fn sub_continuous() {
        let damlev = DamerauLevenshtein::new();
        let sample = [
            (1.0, Text::from_str("abcd"), Text::from_str("xbcd")),
            (2.0, Text::from_str("abcd"), Text::from_str("xycd")),
            (3.0, Text::from_str("abcd"), Text::from_str("xyzd")),

            (1.0, Text::from_str("abcd"), Text::from_str("axcd")),
            (2.0, Text::from_str("abcd"), Text::from_str("axyd")),

            (1.0, Text::from_str("abcd"), Text::from_str("abcz")),
            (2.0, Text::from_str("abcd"), Text::from_str("abyz")),
            (3.0, Text::from_str("abcd"), Text::from_str("axyz")),
        ];
        for (d, t1, t2) in sample.iter() {
            assert_eq!(damlev.distance(&t1.view(0), &t2.view(0)), *d);
        }
    }

    #[test]
    fn trans_continuous() {
        let damlev = DamerauLevenshtein::new();
        let sample = [
            (0.5, Text::from_str("abcd"), Text::from_str("bacd")), // swap 1 and 2
            (1.0, Text::from_str("abcd"), Text::from_str("badc")), // swap 3 and 4
        ];
        for (d, t1, t2) in sample.iter() {
            assert_eq!(damlev.distance(&t1.view(0), &t2.view(0)), *d);
            assert_eq!(damlev.distance(&t2.view(0), &t1.view(0)), *d);
        }
    }

    #[test]
    fn add_del_intermittent() {
        let damlev = DamerauLevenshtein::new();
        let sample = [
            (1.0, Text::from_str("abc"), Text::from_str("xabc")),
            (2.0, Text::from_str("abc"), Text::from_str("xaybc")),
            (3.0, Text::from_str("abc"), Text::from_str("xaybzc")),

            (1.0, Text::from_str("abc"), Text::from_str("abcz")),
            (2.0, Text::from_str("abc"), Text::from_str("abycz")),
            (3.0, Text::from_str("abc"), Text::from_str("axbycz")),
        ];
        for (d, t1, t2) in sample.iter() {
            assert_eq!(damlev.distance(&t1.view(0), &t2.view(0)), *d);
            assert_eq!(damlev.distance(&t2.view(0), &t1.view(0)), *d);
        }
    }

    #[test]
    fn sub_intermittent() {
        let damlev = DamerauLevenshtein::new();
        let sample = [
            (1.0, Text::from_str("abcd"), Text::from_str("xbcd")),
            (2.0, Text::from_str("abcd"), Text::from_str("xbyd")),

            (1.0, Text::from_str("abcd"), Text::from_str("abcy")),
            (2.0, Text::from_str("abcd"), Text::from_str("axcy")),
        ];
        for (d, t1, t2) in sample.iter() {
            assert_eq!(damlev.distance(&t1.view(0), &t2.view(0)), *d);
            assert_eq!(damlev.distance(&t2.view(0), &t1.view(0)), *d);
        }
    }

    #[test]
    fn growth() {
        let damlev = DamerauLevenshtein::new();
        for len in (1..501).step_by(100) {
            let mut s1 = String::with_capacity(len);
            let mut s2 = String::with_capacity(len);
            for _ in 0..len { s1.push('k'); }
            for _ in 0..len { s2.push('l'); }
            let t0 = Text::from_str("");
            let t1 = Text::from_str(&s1);
            let t2 = Text::from_str(&s2);
            assert_eq!(damlev.distance(&t1.view(0), &t1.view(0)), 0.0);
            assert_eq!(damlev.distance(&t1.view(0), &t0.view(0)), len as f64);
            assert_eq!(damlev.distance(&t1.view(0), &t2.view(0)), len as f64);
        }
    }

    #[test]
    pub fn add_del_lang_consonant() {
        let lang   = lang_english();
        let damlev = DamerauLevenshtein::new();
        let text   = |s| Text::from_str(s).set_char_classes(&lang);
        let sample = [
            (1.0, text("pink"), text("spink")),
            (2.0, text("pink"), text("shpink")),
            (3.0, text("pink"), text("schpink")),

            (1.0, text("pink"), text("plink")),
            (2.0, text("pink"), text("prlink")),

            (1.0, text("pink"), text("pinks")),
            (2.0, text("pink"), text("pinkst")),
            (3.0, text("pink"), text("pinkstr")),
        ];
        for (d, t1, t2) in sample.iter() {
            assert_eq!(damlev.distance(&t1.view(0), &t2.view(0)), *d);
            assert_eq!(damlev.distance(&t2.view(0), &t1.view(0)), *d);
        }
    }

    #[test]
    pub fn add_del_lang_vowel() {
        let lang   = lang_english();
        let damlev = DamerauLevenshtein::new();
        let text   = |s| Text::from_str(s).set_char_classes(&lang);
        let sample = [
            (0.5, text("pink"), text("opink")),
            (1.0, text("pink"), text("aopink")),
            (1.5, text("pink"), text("aiopink")),

            (0.5, text("pink"), text("poink")),
            (1.0, text("pink"), text("paoink")),

            (0.5, text("pink"), text("pinky")),
            (1.0, text("pink"), text("pinkie")),
            (1.5, text("pink"), text("pinkaio")),
        ];
        for (d, t1, t2) in sample.iter() {
            assert_eq!(damlev.distance(&t1.view(0), &t2.view(0)), *d);
            assert_eq!(damlev.distance(&t2.view(0), &t1.view(0)), *d);
        }
    }

    #[test]
    fn sub_lang() {
        let lang   = lang_english();
        let damlev = DamerauLevenshtein::new();
        let text   = |s| Text::from_str(s).set_char_classes(&lang);
        let sample = [
            (0.5, text("pinky"), text("punky")),
            (1.0, text("pinky"), text("psnky")),
            (1.0, text("pinky"), text("ponko")),
            (1.5, text("pinky"), text("pesky")),
            (2.0, text("pinky"), text("psnkn")),
        ];
        for (d, t1, t2) in sample.iter() {
            assert_eq!(damlev.distance(&t1.view(0), &t2.view(0)), *d);
            assert_eq!(damlev.distance(&t2.view(0), &t1.view(0)), *d);
        }
    }

    #[test]
    fn add_del_double_lang() {
        let lang   = lang_english();
        let damlev = DamerauLevenshtein::new();
        let text   = |s| Text::from_str(s).set_char_classes(&lang);
        let sample = [
            // vowel end
            (0.5, text("pink"),  text("pinky")),
            (1.0, text("pink"),  text("pinkyy")),
            // vowel mid
            (0.5, text("pinky"),  text("piinky")),
            (1.0, text("pinky"),  text("piiinky")),
            // consonant mid
            (1.5, text("pinky"),  text("pinkxxy")),
            (2.0, text("pinky"),  text("pinkxxxy")),
            // consonant end
            (1.5, text("pinky"),  text("pinkyss")),
            (2.0, text("pinky"),  text("pinkysss")),
        ];
        for (d, t1, t2) in sample.iter() {
            assert_eq!(damlev.distance(&t1.view(0), &t2.view(0)), *d);
            assert_eq!(damlev.distance(&t2.view(0), &t1.view(0)), *d);
        }
    }

    #[test]
    fn sub_double_lang() {
        let lang   = lang_english();
        let damlev = DamerauLevenshtein::new();
        let text   = |s| Text::from_str(s).set_char_classes(&lang);
        let sample = [
            // consonant mid
            (0.5, text("pinky"), text("pinkky")),
            (1.0, text("pinky"), text("pinkkky")),
            (1.5, text("pinky"), text("pinkkkky")),
            // vowel end
            (0.5, text("pinky"), text("pinkyy")),
            (1.0, text("pinky"), text("pinkyyy")),
            (1.5, text("pinky"), text("pinkyyyy")),
            // consonant begin
            (0.5, text("pinky"), text("ppinky")),
            (1.0, text("pinky"), text("pppinky")),
            (1.5, text("pinky"), text("ppppinky")),
            // consonant mid
            (1.5, text("pinky"), text("pillky")),
            (2.0, text("pinky"), text("pilllky")),
            // vowel mid
            (1.0, text("pinky"), text("poonky")),
            (1.5, text("pinky"), text("pooonky")),
        ];
        for (d, t1, t2) in sample.iter() {
            assert_eq!(damlev.distance(&t1.view(0), &t2.view(0)), *d);
            assert_eq!(damlev.distance(&t2.view(0), &t1.view(0)), *d);
        }
    }

    #[test]
    fn notalpha_lang() {
        let lang   = lang_english();
        let damlev = DamerauLevenshtein::new();
        let text   = |s| Text::from_str(s).set_char_classes(&lang);
        let sample = [
            (0.5, text("pinky"), text("p_nky")),
            (1.0, text("pinky"), text("p_nk_")),
            (2.0, text("pinky"), text("__nk_")),
        ];
        for (d, t1, t2) in sample.iter() {
            assert_eq!(damlev.distance(&t1.view(0), &t2.view(0)), *d);
            assert_eq!(damlev.distance(&t2.view(0), &t1.view(0)), *d);
        }
    }
}
