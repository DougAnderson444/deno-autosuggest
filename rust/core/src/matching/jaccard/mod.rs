use std::cell::RefCell;
use std::hash::Hash;
use std::collections::HashSet;


const DEFAULT_CAPACITY: usize = 20;


pub struct Jaccard<T: PartialEq + Copy + Ord + Hash> {
    set1: RefCell<HashSet<T>>,
    set2: RefCell<HashSet<T>>,
}


impl<T: PartialEq + Copy + Ord + Hash> Jaccard<T> {
    pub fn new() -> Self {
        Self {
            set1: RefCell::new(HashSet::with_capacity(DEFAULT_CAPACITY)),
            set2: RefCell::new(HashSet::with_capacity(DEFAULT_CAPACITY)),
        }
    }

    pub fn similarity(&self, slice1: &[T], slice2: &[T]) -> f64 {
        match (slice1.len(), slice2.len()) {
            (0, 0) => return 1.0,
            (0, _) => return 0.0,
            (_, 0) => return 0.0,
            (_, _) => { },
        }

        let set1 = &mut *self.set1.borrow_mut();
        let set2 = &mut *self.set2.borrow_mut();

        set1.clear();
        set2.clear();
        for &item1 in slice1 { set1.insert(item1); }
        for &item2 in slice2 { set2.insert(item2); }

        let count_union        = set1.union(set2).count();
        let count_intersection = set1.intersection(set2).count();

        count_intersection as f64 / count_union as f64
    }

    pub fn rel_dist(&self, slice1: &[T], slice2: &[T]) -> f64 {
        1.0 - self.similarity(slice1, slice2)
    }
}


#[cfg(test)]
mod tests {
    use super::Jaccard;


    fn round2(x: f64) -> f64 {
        (x * 100.0).round() / 100.0
    }


    #[test]
    fn empty_both() {
        let jaccard = Jaccard::new();
        let s1: &[usize] = &[];
        let s2: &[usize] = &[];
        assert_eq!(round2(jaccard.rel_dist(s1, s2)), 0.0);
    }


    #[test]
    fn empty_one() {
        let jaccard = Jaccard::new();
        let sample = [
            (1.00, vec![], vec![1]),
            (1.00, vec![], vec![1, 2]),
            (1.00, vec![], vec![1, 2, 3]),
            (1.00, vec![], vec![1, 2, 3, 4]),
        ];
        for (d, s1, s2) in sample.iter() {
            assert_eq!(round2(jaccard.rel_dist(s1, s2)), *d);
        }
    }


    #[test]
    fn equal() {
        let jaccard = Jaccard::new();
        let sample = [
            vec![1],
            vec![1, 2],
            vec![1, 2, 3],
            vec![1, 2, 3, 4],
        ];
        for s in sample.iter() {
            assert_eq!(round2(jaccard.rel_dist(s, s)), 0.0);
        }
    }


    #[test]
    fn different() {
        let jaccard = Jaccard::new();
        let sample = [
            (1.00, vec![1],          vec![5]),
            (1.00, vec![1, 2],       vec![5, 6]),
            (1.00, vec![1, 2, 3],    vec![5, 6, 7]),
            (1.00, vec![1, 2, 3, 4], vec![5, 6, 7, 8]),
        ];
        for (d, s1, s2) in sample.iter() {
            assert_eq!(round2(jaccard.rel_dist(s1, s2)), *d);
        }
    }


    #[test]
    fn partial() {
        let jaccard = Jaccard::new();
        let sample = [
            (0.25, vec![1, 2, 3, 4], vec![1, 2, 3]),
            (0.50, vec![1, 2, 3, 4], vec![1, 2]),
            (0.75, vec![1, 2, 3, 4], vec![1]),
        ];
        for (d, s1, s2) in sample.iter() {
            assert_eq!(round2(jaccard.rel_dist(s1, s2)), *d);
        }
    }


    #[test]
    fn permutated() {
        let jaccard = Jaccard::new();
        let sample = [
            (0.00, vec![1, 2, 3, 4], vec![3, 1, 2, 4]),
            (0.00, vec![1, 2, 3, 4], vec![4, 2, 3, 1]),
            (0.00, vec![1, 2, 3, 4], vec![1, 4, 2 ,3]),
        ];
        for (d, s1, s2) in sample.iter() {
            assert_eq!(round2(jaccard.rel_dist(s1, s2)), *d);
        }
    }


    #[test]
    fn repeated() {
        let jaccard = Jaccard::new();
        let sample = [
            (0.75, vec![1, 2, 3, 4], vec![1, 1, 1, 1]),
            (0.50, vec![1, 2, 3, 4], vec![2, 1, 2, 1]),
            (0.25, vec![1, 2, 3, 4], vec![3, 2, 1, 3]),
        ];
        for (d, s1, s2) in sample.iter() {
            assert_eq!(round2(jaccard.rel_dist(s1, s2)), *d);
        }
    }


    #[test]
    fn complex() {
        let jaccard = Jaccard::new();
        let sample = [
            (0.40, vec![1, 2, 3, 4], vec![0, 1, 2, 4]),
            (0.40, vec![1, 2, 3, 4], vec![0, 2, 3, 1]),
            (0.50, vec![1, 2, 3, 4], vec![0, 4, 2 ,3, 5]),
        ];
        for (d, s1, s2) in sample.iter() {
            assert_eq!(round2(jaccard.rel_dist(s1, s2)), *d);
        }
    }
}