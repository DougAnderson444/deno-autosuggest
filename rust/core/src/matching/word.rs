use crate::tokenization::Word;
use super::WordMatch;
use super::damlev::DamerauLevenshtein;
use super::jaccard::Jaccard;


const LENGTH_THRESHOLD:  f64 = 0.26;
const JACCARD_THRESHOLD: f64 = 0.41;
const DAMLEV_THRESHOLD:  f64 = 0.26;

thread_local! {
    static DAMLEV:  DamerauLevenshtein = DamerauLevenshtein::new();
    static JACCARD: Jaccard<char>      = Jaccard::new();
}


pub fn word_match(rword: &Word, qword: &Word, rchars: &[char], qchars: &[char]) -> Option<WordMatch> {
    if qword.is_empty() || rword.is_empty() {
        return None;
    }
    if !length_check(rword, qword) {
        return None;
    }
    if !jaccard_check(rword, qword, rchars, qchars) {
        return None;
    }

    let rchars = rword.view(rchars);
    let qchars = qword.view(qchars);
    let mut best_match: Option<WordMatch> = None;

    DAMLEV.with(|damlev| {
        damlev.distance(qchars, rchars);
        let dists = &*damlev.dists.borrow();

        let left  = if qword.fin { max!(qword.stem, rword.stem) } else { qword.stem } - 1;
        let right = max!(qword.len(), rword.len()) + 1;

        if right <= left { return; }

        let range = (left .. right).rev();

        for rlen in range.clone() {
            for qlen in range.clone() {
                // Out of bounds.
                if qlen > qword.len() { continue; }
                if rlen > rword.len() { continue; }
                // Left margin is for insertion/deletion, not for both prefixes at the same time.
                if rlen == left && qlen == left  { continue; }
                // Compare full words only if query is finished.
                if qword.fin && rlen < rword.stem { break; }
                if qword.fin && qlen < qword.stem { break; }
                // Words with 2+ insertions/deletions are mismatched by default.
                if (qlen as isize - rlen as isize).abs() > 1 { continue; }

                let dist = dists.get(qlen + 1, rlen + 1);
                let rel  = dist as f64 / max!(qlen, rlen, 1) as f64;

                if rel > DAMLEV_THRESHOLD { continue; }

                best_match = best_match
                    .take()
                    .filter(|m| m.typos <= dist)
                    .or_else(|| Some(WordMatch::new(
                        qword,
                        rword,
                        qlen,
                        rlen,
                        dist,
                    )));

                if dist == 0 {
                    break;
                }
            }
        }
    });

    best_match
}


pub fn length_check(rword: &Word, qword: &Word) -> bool {
    let qlen  = qword.len();
    let rlen  = if qword.fin { rword.len() } else { min!(qlen, rword.len()) };

    if qlen <= 1 || rlen <= 1 {
        return qlen == rlen;
    }

    let long  = max!(qlen, rlen);
    let short = min!(qlen, rlen);
    let dist  = 1.0 - (short as f64 / long as f64);

    dist < LENGTH_THRESHOLD
}


pub fn jaccard_check(rword: &Word, qword: &Word, rchars: &[char], qchars: &[char]) -> bool {
    let rchars = rword.view(rchars);
    let qchars = qword.view(qchars);
    let rslice = if qword.fin { &rchars } else { &rchars[.. min!(qword.len(), rword.len())] };
    let dist   = JACCARD.with(|j| j.rel_dist(&rslice, &qchars));
    dist < JACCARD_THRESHOLD
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use crate::utils::to_vec;
    use crate::tokenization::{Chars, Word, Text};
    use crate::lang::lang_english;
    use super::{length_check, jaccard_check, word_match};


    fn text(s: &str) -> Text<Vec<char>> {
        Text::from_str(s).split(&Chars::Whitespaces)
    }

    #[test]
    fn length_check_len_4() {
        let sample = [
            (false, "m"),
            (false, "ma"),
            (true,  "mai"),
            (true,  "mail"),
            (true,  "mailb"),
            (false, "mailbo"),
            (false, "mailbox"),
        ];
        for &(expect, query) in sample.iter() {
            let rtext  = text("mail");
            let qtext  = text(query);
            let result = length_check(&rtext.words[0], &qtext.words[0]);
            assert_eq!(result, expect, "Failed length_check(\"mail\", \"{}\") == {}", query, expect);
        }
    }

    #[test]
    fn length_check_len_7() {
        let sample = [
            (false, "m"),
            (false, "ma"),
            (false, "mai"),
            (false, "mail"),
            (false, "mailb"),
            (true,  "mailbo"),
            (true,  "mailbox"),
            (true,  "mailboxe"),
            (true,  "mailboxes"),
            (false, "mailboxess"),
        ];
        for &(expect, query) in sample.iter() {
            let rtext  = text("mailbox");
            let qtext  = text(query);
            let result = length_check(&rtext.words[0], &qtext.words[0]);
            assert_eq!(result, expect, "Failed length_check(\"mailbox\", \"{}\") == {}", query, expect);
        }
    }

    #[test]
    fn jaccard_check_len_4() {
        let sample = [
            (true,  "mail"),
            (true,  "bail"),
            (false, "bait"),
            (false, "balt"),
            (false, "bolt"),
        ];
        for &(expect, query) in sample.iter() {
            let rtext  = text("mail");
            let qtext  = text(query);
            let result = jaccard_check(&rtext.words[0], &qtext.words[0], &rtext.chars, &qtext.chars);
            assert_eq!(result, expect, "Failed jaccard_check(\"mail\", \"{}\") == {}", query, expect);
        }
    }

    #[test]
    fn jaccard_check_len_7() {
        let sample = [
            (true,  "mailbox"),
            (true,  "mailbot"),
            (false, "railbot"),
            (false, "raidbot"),
            (false, "roidbot"),
        ];
        for &(expect, query) in sample.iter() {
            let rtext  = text("mailbox");
            let qtext  = text(query);
            let result = jaccard_check(&rtext.words[0], &qtext.words[0], &rtext.chars, &qtext.chars);
            assert_eq!(result, expect, "Failed jaccard_check(\"mailbox\", \"{}\") == {}", query, expect);
        }
    }

    #[test]
    fn jaccard_check_len_7_reduction() {
        let sample = [
            (true,  "mailbox"),
            (true,  "mailbxx"),
            (true,  "mailxxx"),
            (false, "maixxxx"),
            (false, "maxxxxx"),
        ];
        for &(expect, query) in sample.iter() {
            let rtext  = text("mailbox");
            let qtext  = text(query);
            let result = jaccard_check(&rtext.words[0], &qtext.words[0], &rtext.chars, &qtext.chars);
            assert_eq!(result, expect, "Failed jaccard_check(\"mailbox\", \"{}\") == {}", query, expect);
        }
    }

    #[test]
    fn jaccard_check_unfinished() {
        let sample = [
            (true,  "m"),
            (true,  "ma"),
            (true,  "mai"),
            (true,  "mail"),
            (true,  "mailb"),
            (true,  "mailbo"),
            (false, "mailbox"),
        ];
        for &(expect, query) in sample.iter() {
            let rtext  = text("mail");
            let qtext  = text(query).fin(false);
            let result = jaccard_check(&rtext.words[0], &qtext.words[0], &rtext.chars, &qtext.chars);
            assert_eq!(result, expect, "Failed jaccard_check(\"mail\", \"{}\") == {}", query, expect);
        }
    }

    #[test]
    fn length_check_unfinished() {
        let sample = [
            (true,  "m"),
            (true,  "ma"),
            (true,  "mai"),
            (true,  "mail"),
            (true,  "mailb"),
            (false, "mailbo"),
            (false, "mailbox"),
        ];
        for &(expect, query) in sample.iter() {
            let rtext  = text("mail");
            let qtext  = text(query).fin(false);
            let result = length_check(&rtext.words[0], &qtext.words[0]);
            assert_eq!(result, expect, "Failed length_check(\"mail\", \"{}\") == {}", query, expect);
        }
    }

    // Match word
    // ----------------------------------------------------------------

    #[test]
    fn match_word_empty_both() {
        let qchars = to_vec("");
        let rchars = to_vec("");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_eq!(word_match(&rword, &qword, &rchars[..], &qchars[..]), None);
    }


    #[test]
    fn match_word_empty_record() {
        let qchars = to_vec("mailbox");
        let rchars = to_vec("");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_eq!(word_match(&rword, &qword, &rchars[..], &qchars[..]), None);
    }


    #[test]
    fn match_word_empty_query() {
        let qchars = to_vec("");
        let rchars = to_vec("mailbox");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_eq!(word_match(&rword, &qword, &rchars[..], &qchars[..]), None);
    }


    // Match word: prefixed by same length word
    // ----------------------------------------------------------------

    #[test]
    fn match_word_full_strict() {
        let qchars = to_vec("mailbox");
        let rchars = to_vec("mailbox");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_full_fuzzy_insertion() {
        let qchars = to_vec("mailybox");
        let rchars = to_vec("mailbox");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_full_fuzzy_deletion() {
        let qchars = to_vec("mailox");
        let rchars = to_vec("mailbox");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_full_fuzzy_transposition() {
        let qchars = to_vec("maiblox");
        let rchars = to_vec("mailbox");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_full_query_too_long() {
        let qchars1 = to_vec("mailboxes");
        let qchars2 = to_vec("mailboxes");
        let rchars  = to_vec("mail");
        let qword1  = Word::new(qchars1.len()).fin(true);
        let qword2  = Word::new(qchars2.len()).fin(false);
        let rword   = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword1, &rchars[..], &qchars1[..]));
        assert_debug_snapshot!(word_match(&rword, &qword2, &rchars[..], &qchars2[..]));
    }

    #[test]
    fn match_word_full_stem() {
        let     rchars  = to_vec("universe");
        let     qchars1 = to_vec("university");
        let     qchars2 = to_vec("university");
        let mut rword   = Word::new(rchars.len());
        let mut qword1  = Word::new(qchars1.len());
        let     qword2  = Word::new(qchars2.len());

        let lang = lang_english();
        qword1.stem(&qchars1[..], &lang);
        rword .stem(&rchars[..],  &lang);

        assert_debug_snapshot!(word_match(&rword, &qword1, &rchars[..], &qchars1[..]));
        assert_debug_snapshot!(word_match(&rword, &qword2, &rchars[..], &qchars2[..]));
    }


    // Match word: prefixed by lesser length word
    // ----------------------------------------------------------------

    #[test]
    fn match_word_partial_strict() {
        let qchars = to_vec("mailb");
        let rchars = to_vec("mailbox");
        let qword = Word::new(qchars.len()).fin(false);
        let rword = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_partial_fuzzy_insertion() {
        let qchars = to_vec("maiylb");
        let rchars = to_vec("mailbox");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_partial_fuzzy_deletion() {
        let qchars = to_vec("maib");
        let rchars = to_vec("mailbox");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_partial_fuzzy_transposition() {
        let qchars = to_vec("malib");
        let rchars = to_vec("mailbox");
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }
}
