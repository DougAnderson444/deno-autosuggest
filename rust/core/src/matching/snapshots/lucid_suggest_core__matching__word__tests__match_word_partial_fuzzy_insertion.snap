---
source: src/matching/word.rs
expression: "word_match(&rword, &qword, &rchars[..], &qchars[..])"
---
Some(
    WordMatch {
        query: MatchSide {
            ix: 0,
            len: 6,
            slice: (
                0,
                6,
            ),
            function: false,
        },
        record: MatchSide {
            ix: 0,
            len: 7,
            slice: (
                0,
                5,
            ),
            function: false,
        },
        typos: 1,
        fin: false,
    },
)
